### PR TITLE
Validation repo (EXPOSUREAPP-7618)

### PR DESCRIPTION
**Pull request by chris-cwa**
Implementation of the Validation rule API, rule Server and caching.

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/DccValidationCache.kt [hunk 1]
```diff
@@ -13,31 +14,45 @@ class DccValidationCache @Inject constructor(
...
+        invalidationRulesCacheFile.save(data, "raw invalidation rules")
+
+    @VisibleForTesting
+    internal suspend fun File.load(descr: String): String? = mutex.withLock {
         try {
-            if (countryCacheFile.exists()) countryCacheFile.readText() else null
+            if (exists()) readText() else null
         } catch (e: Exception) {
-            Timber.tag(TAG).e(e, "Failed to load raw dcc countries from cache.")
+            Timber.tag(TAG).e(e, "Failed to load $descr from cache.")
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/DccValidationCache.kt:38]: Alternatively, use the filename to differentiate log messages.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/rule/DccValidationRuleSet.kt [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package de.rki.coronawarnapp.covidcertificate.validation.core.rule
+
+import android.os.Parcelable
+import kotlinx.parcelize.Parcelize
+
+@Parcelize
+data class DccValidationRuleSet(
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/rule/DccValidationRuleSet.kt:7]: Where is this used?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/server/DccValidationServer.kt [hunk 3]
```diff
@@ -24,17 +30,27 @@ class DccValidationServer @Inject constructor(
...
 
     private val dccValidationRuleApi: DccValidationRuleApi
         get() = rulesApi.get()
 
-    suspend fun ruleSet(ruleTypeDcc: DccValidationRule.Type): Set<DccValidationRule> {
-        return when (ruleTypeDcc) {
-            DccValidationRule.Type.ACCEPTANCE -> dccValidationRuleApi.acceptanceRules()
-            DccValidationRule.Type.INVALIDATION -> dccValidationRuleApi.invalidationRules()
+    suspend fun ruleSetJson(ruleTypeDcc: DccValidationRule.Type): String =
+        withContext(dispatcherProvider.Default) {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/server/DccValidationServer.kt:40]: We shouldn't explicitly switch to the `Default`, either no switch (and let the caller decide) or switch to `IO` (probably best). Due to the amount of threads in `Default` being capped, a http request timing out could block a thread for 10-20 seconds potentially starving other coroutines.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/server/DccValidationServer.kt [hunk 4]
```diff
@@ -81,6 +97,29 @@ class DccValidationServer @Inject constructor(
         }
     }
 
+    private fun ResponseBody.parseBody() = parseBody(byteStream())
+
+    @VisibleForTesting
+    internal fun parseBody(inputStream: InputStream): String {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/server/DccValidationServer.kt:103]: Can parseBody also be used for the countries?

**Review by d4rken — COMMENTED** [review]
Thanks for the refactoring!

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/rule/DccValidationRuleSet.kt (hunk 2 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/rule/DccValidationRuleSet.kt:7]: it is not :)

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/server/DccValidationServer.kt (hunk 4 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/validation/core/server/DccValidationServer.kt:103]: there sure is a lot of duplicate code!

**Review by chiljamgossow — DISMISSED** [review]
Merge and test?

**Review by d4rken — DISMISSED** [review]

**Review by d4rken — APPROVED** [review]

**Review by chiljamgossow — APPROVED** [review]
Looking good

**Pull request merged**